### PR TITLE
Enhance/parallel linkstatus

### PR DESCRIPTION
**Pull request by rueyjye**
Separate parallel for e2e and linkstatus.

Change parallel helper into module.

### CODE REVIEW

File: lib/bucky/core/test_core/test_manager.rb [hunk 1]
```diff
@@ -8,7 +8,51 @@
...
+          end
+          Process.waitall
+        end
+
+        def parallel_distribute_into_workers(data_set, max_processes, &block)
+          # For checking on linkstatus
+          divisor_of_data = max_processes
+          num_of_works_in_pre_worker = (data_set.length.to_f / divisor_of_data.to_f).ceil
+          # Slice data_set into few parts that depends on workers
+          data_set.each_slice(num_of_works_in_pre_worker) do |data_for_pre_worker|
```
Comment (NA070) [inline lib/bucky/core/test_core/test_manager.rb]: I think there are times when this implementation doesn't work fine.
```
max =4
data_set = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

(data_set.length.to_f / 4.to_f).ceil
# => 3

data_set.each_slice(3) {|d| p d }
# =>[1, 2, 3]
# =>[4, 5, 6]
# =>[7, 8, 9]
# =>[10]

# Works fine
# =>[1, 2, 3]
# =>[4, 5, 6]
# =>[7, 8]
# =>[9, 10]
```

How about implementation like this?
```
max = 4
data_set = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

devided_data_set = Array.new
max.times {|i| devided_data_set << [] }
data_index = 0
data_set.each do |d|
  data_index = 0 if data_index >= max
  devided_data_set[data_index] << d
  data_index +=1
end
p devided_data_set

# => [[1, 5, 9], [2, 6, 10], [3, 7], [4, 8]]
```

File: lib/bucky/core/test_core/test_manager.rb [hunk 2]
```diff
@@ -45,8 +89,16 @@ def load_test_suites
 
         # Generate and execute test
         def do_test_suites(test_suite_data)
-          parallel_num = Bucky::Utils::Config.instance[:parallel_num]
-          parallel_helper(test_suite_data, parallel_num)
+          # For checking on linkstatus
+          link_status_url_log = {}
```
Comment (NA070) [inline lib/bucky/core/test_core/test_manager.rb]: I think unnecessary variables are defined.
When test category is e2e, variables for linkstatus are unnecessary.
How about you?

File: spec/core/test_core/test_manager_spec.rb [hunk 3]
```diff
@@ -95,11 +97,38 @@
...
 
-      it 'create test class instance in fork' do
-        allow(tm).to receive(:fork) do |&block|
-          expect(block.call).to eq(Bucky::Core::TestCore::TestClasses::TestSpecPcE2etest)
+      it 'call target block in fork' do
+        allow(parallel_helper).to receive(:fork) do |&block|
+          expect(block.call).to eq('Block mock')
         end
-        tm.send(:parallel_helper, test_suite_data, parallel_num)
+        parallel_helper.send(:parallel_new_worker_each, test_suite_data_e2e, e2e_parallel_num, &block_mock)
```
Comment (NA070) [inline spec/core/test_core/test_manager_spec.rb]: Could tell me why you use .send?

File: spec/core/test_core/test_manager_spec.rb [hunk 4]
```diff
@@ -95,11 +97,38 @@
...
+            } }
+        ]
+      end
+      it 'make workers in fork then call block in each' do
+        allow(parallel_helper).to receive(:fork) do |&block_fork|
+          allow(block_fork.call).to receive(:each) do |&block_each|
+            expect(block_each.call).to eq('Block mock')
+          end
+        end
+        parallel_helper.send(:parallel_distribute_into_workers, test_suite_data_linkstatus, linkstatus_parallel_num, &block_mock)
```
Comment (NA070) [inline spec/core/test_core/test_manager_spec.rb]: Could tell me why you use .send?

File: lib/bucky/core/test_core/test_manager.rb (hunk 1 above)
Comment (rueyjye) [reply lib/bucky/core/test_core/test_manager.rb]: Thanks of your advise.
I want to change it like bellow.
```
max =4
data_set = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

data_set_grouped = data_set.group_by.with_index { |elem,index| index%max }
p data_set_grouped.values

# =>  [[1, 5, 9], [2, 6, 10], [3, 7], [4, 8]]
```

File: spec/core/test_core/test_manager_spec.rb (hunk 3 above)
Comment (rueyjye) [reply spec/core/test_core/test_manager_spec.rb]: Because it's a test for private method.

File: spec/core/test_core/test_manager_spec.rb (hunk 4 above)
Comment (rueyjye) [reply spec/core/test_core/test_manager_spec.rb]: Because it's a test for private method.

File: lib/bucky/core/test_core/test_manager.rb (hunk 2 above)
Comment (rueyjye) [reply lib/bucky/core/test_core/test_manager.rb]: I'll fix it!

**Pull request merged**